Add support for decorating Pull Requests (Gitlab Server)

tisoft opened the pull request:
Based on the work of @4n4n4s in #31 I've implemented the pull request decoration for gitlab. It's work in progress currently. Output can be seen here: https://gitlab.com/tisoft/sonarqube-community-branch-plugin-gitlab-test/merge_requests/1

Current features:
* set Commit Status to succes/failed based on sonarqube quality status
* add a summery comment
* add line comments to files (still a bit buggy)
* Deletion of old comments (configurable)

I will try to switch to conversations instead of notes/comments and see how they behave. Maybe both can be supported. Still playing around with this. :smile:

hubermi commented:
I'd like to contribute to this PR. Gitlab CI has the ability to show a code quality report right in the merge request widget area (see https://docs.gitlab.com/ee/user/project/merge_requests/code_quality.html#example-configuration)

Currently only reports from code climate analysis are supported (and I guess that gitlab will stick to this format only) so we need some code that transforms the sonar result to the code climate format. 

You can then add that file as report artifact.

If you don't mind, I'd like to add this functionality.

mfolnovic commented:
I think this is great to start with, I would probably implement deletion of old comments just because (as you can see from your examples), you'll get a lot of comments. :) 

One thing I definitely miss is opening and closing discussions based on new sonar smells, which makes code review a lot easier. You could then:
- see those comments during scrolling through all changes (which you can't see with just commit comments)
- you can directly discuss if it's false positive / won't fix or if it should be fixed

But there are a bunch of corner cases so I'd rather do that separately. :)

tisoft commented:
Deletion of old comments is already implemented. Will look into "real discussions" next week.

mc1arke commented:
@tisoft Could you rebase this against the updated base branch please?

tisoft commented:
Merged in the new changes. Compiles and runs again.

I changed the visibility of `findQualityGateCondition` in `AnalysisDetails` to public, to get the new code coverage, since gitlab allows to submit that in commit status posts.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/AnalysisDetails.java:
```diff
@@ -170,6 +173,39 @@ public String createAnalysisSummary(FormatterFactory formatterFactory) {
+        String resolution = issue.resolution();
+        if (StringUtils.isNotBlank(resolution)) {
+            document.addChild(new Paragraph(new Text(String.format("Resolution: %s ", resolution))));
+        }
```
Making the `addChild` method on `Node` doesn't seem right as it introduces mutability into a previously immutable construct. I'd suggest using an empty `Text` node during construction if you decide you don't need a real element:
```java
        Long effort = issue.effortInMinutes();
        Node effortNode = (null == effort ? new Text("") : new Paragraph(new Text(String.format("Duration (min): %s", effort))));
        
        String resolution = issue.resolution();
        Node resolutionNode = (StringUtils.isBlank(resolution) ? new Text("") : new Paragraph(new Text(String.format("Resolution: %s ", resolution))));
        
        Document document = new Document(
                new Paragraph(new Text(String.format("Type: %s", issue.type().name()))),
                new Paragraph(new Image(issue.severity(), String.format("%s/checks/Severity/%s.svg?sanitize=true", baseImageUrl, issue.severity().toLowerCase())), new Text(issue.severity())),
                new Paragraph(new Text(String.format("Message: %s", issue.getMessage()))),
                effortNode,
                resolutionNode
            );
```

tisoft replied:
This was taken from the bitbucket implementation, so that we have a consistent formatting of Issues in line comments.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/AnalysisDetails.java:
```diff
@@ -170,6 +173,39 @@ public String createAnalysisSummary(FormatterFactory formatterFactory) {
+        return formatterFactory.documentFormatter().format(document, formatterFactory);
+    }
+
+    public String getSCMPathForIssue(PostAnalysisIssueVisitor.ComponentIssue componentIssue) {
```
This should return an `Optional<String>` rather than an empty String on no result

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/GitlabServerPullRequestDecorator.java:
```diff
@@ -0,0 +1,379 @@
+/*
+ * Copyright (C) 2019 Michael Clarke
```
Do you not want to put your own name in here?

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/GitlabServerPullRequestDecorator.java:
```diff
@@ -0,0 +1,379 @@
+                        StandardCharsets.UTF_8.name()), URLEncoder
+                        .encode(analysis.getBranchName(),
+                                StandardCharsets.UTF_8.name())), StandardCharsets.UTF_8.name());
+        //statusPostUrl+="&description="+URLEncoder.encode(comment, StandardCharsets.UTF_8.name());
```
Please don't leave commented out code 😄

tisoft replied:
Still experimenting with the descrition. will either enable it or remove the line

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/GitlabServerPullRequestDecorator.java:
```diff
@@ -0,0 +1,379 @@
+                                StandardCharsets.UTF_8.name())), StandardCharsets.UTF_8.name());
+        //statusPostUrl+="&description="+URLEncoder.encode(comment, StandardCharsets.UTF_8.name());
+        statusPostUrl+="&coverage="+coverage;
+        //TODO: add pipelineId if we have it
```
Is this TODO still needed, and if so, when does this need done?

tisoft replied:
Need to find out how I can get the environment variable `CI_PIPELINE_ID` from the scanner side. When the build is run with gilab ci and there are multiple pipelines for a MR, the result will then be attached to the correct pipeline.

tisoft replied:
@mc1arke Do you have an idea how to get the value of an environment variable from the scanner back to the sonar server? Gitlab CI sets the `CI_PIPELINE_ID` variable and I would like to pass that down to the status posting. Otherwise the wrong pipeline could be annotated with the sonar result.

I couldn't find a way, so if you also don#t know anything, I will just remove the TODO comment.

mc1arke replied:
You won't be able to access the variable directly at this point as this is the Compute Engine executing on the Sonarqube server, rather than the scanner on the Gitlab pipeline. I don't think there would be an easy way to pass this information to the compute engine, but I'll see if there's any way to fudge it.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/GitlabServerPullRequestDecorator.java:
```diff
@@ -0,0 +1,379 @@
+            if (null != httpResponse && httpResponse.toString().contains("Cannot transition status")) {
+                // Workaround for https://gitlab.com/gitlab-org/gitlab-ce/issues/25807
+                LOGGER.debug("Transition status is already {}", status);
+            } else validateGitlabResponse(httpResponse, 201, "Comment posted");
```
Please use braces are all if/else blocks, even if they're just a single line

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/GitlabServerPullRequestDecorator.java:
```diff
@@ -0,0 +1,379 @@
+                String.format("%s must be specified in the project configuration", propertyName)));
+    }
+
+    public String getNextUrl(HttpResponse httpResponse) {
```
Could this be `private static`?

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/GitlabServerPullRequestDecorator.java:
```diff
@@ -0,0 +1,379 @@
+        }
+    }
+
+    protected void postStatus(String statusPostUrl, Map<String, String> headers, AnalysisDetails analysis, String comment, String coverage, boolean sendRequest) throws IOException{
```
Could this be `private`?

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/response/Commit.java:
```diff
@@ -0,0 +1,13 @@
+package com.github.mc1arke.sonarqube.plugin.ce.pullrequest.gitlab.response;
+
+public class Commit {
+    private String id;
```
Make this `final` and initialize it in the constructor. If it's being set by JSON deserialisation, then use `@JsonCreator` and `@JsonProperty`, e.g.
```java
    @JsonCreator
    public Commit(@JsonProperty("id") String id) {
        super();
        this.id = id;
    }
````

tisoft commented:
@mc1arke thanks for the review! Will have a look into the issues next year🎇.

mc1arke commented on README.md:
```diff
@@ -21,4 +21,58 @@ Either build the project or [download a compatible release version of the plugin
 # Features
 The plugin is intended to support the [features and parameters specified in the SonarQube documentation](https://docs.sonarqube.org/latest/branches/overview/), with the following caveats
-* __Pull Requests:__ Analysis of Pull Requests is fully supported, but the decoration of pull requests with any issues is not currently supported
+* __Pull Requests:__ Analysis of Pull Requests is fully supported, but the decoration of pull requests is only currently available for Github, and only as an experimental feature
```
This probably needs to also mention Gitlab

tisoft commented:
- Rebased against your latest base branch
- removed bitbucket stuff
- fixed some review comments

Where can I get the configuration settings, that are used by the "official" plugin, so that I can reuse the configuration names for those?

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/CommunityBranchPlugin.java:
```diff
@@ -67,8 +71,80 @@ public void load(CoreExtension.Context context) {
+                        .defaultValue("https://raw.githubusercontent.com/mc1arke/sonarqube-community-branch-plugin/pr-decoration/src/main/resources/pr-decoration-images")
+                        .description("Default: https://raw.githubusercontent.com/mc1arke/sonarqube-community-branch-plugin/master/src/main/resources/pr-decoration-images")
+                        .type(PropertyType.STRING)
+                        .build(),
```
See https://github.com/mc1arke/sonarqube-community-branch-plugin/pull/31#discussion_r362045510 - you shouldn't need to define this as a property, unless there was something you were specifically trying to do?

tisoft replied:
removed.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/AnalysisDetails.java:
```diff
@@ -170,6 +173,39 @@ public String createAnalysisSummary(FormatterFactory formatterFactory) {
+        String resolution = issue.resolution();
+        if (StringUtils.isNotBlank(resolution)) {
+            document.addChild(new Paragraph(new Text(String.format("Resolution: %s ", resolution))));
+        }*/
```
Don't forget to remove this commented out code

tisoft replied:
replaced with implementation from bitbucket

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/CommunityBranchPlugin.java:
```diff
@@ -67,8 +71,80 @@ public void load(CoreExtension.Context context) {
+                PropertyDefinition.builder("sonar.pullrequest.file.comment.enabled")
+                        .category(PULL_REQUEST_CATEGORY_LABEL)
+                        .subCategory(GENERAL)
+                        .onlyOnQualifiers(Qualifiers.PROJECT)
```
Is this not something that should be global across all SonarQube projects, rather than specific to each project? If so, this should probably be:
```suggestion
                        .onlyOnQualifiers(Qualifiers.APP)
```

tisoft replied:
using the same parameter as the bitbucket integration

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/CommunityBranchPlugin.java:
```diff
@@ -67,8 +71,80 @@ public void load(CoreExtension.Context context) {
+                PropertyDefinition.builder("sonar.pullrequest.gitlab.url")
+                        .category(PULL_REQUEST_CATEGORY_LABEL)
+                        .subCategory(GITLAB_INTEGRATION_SUBCATEGORY_LABEL)
+                        .onQualifiers(Qualifiers.PROJECT)
```
Should this be global across all projects, or set on each project?
```suggestion
                        .onQualifiers(Qualifiers.APP)
```

tisoft replied:
this needs to be per project, since each project might be connected to a different gitlab instance

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/CommunityBranchPlugin.java:
```diff
@@ -67,8 +71,80 @@ public void load(CoreExtension.Context context) {
+                PropertyDefinition.builder("sonar.pullrequest.gitlab.token")
+                        .category(PULL_REQUEST_CATEGORY_LABEL)
+                        .subCategory(GITLAB_INTEGRATION_SUBCATEGORY_LABEL)
+                        .onQualifiers(Qualifiers.PROJECT)
```
Should this be global across all projects, or set on each project?
```suggestion
                        .onQualifiers(Qualifiers.APP)
```

tisoft replied:
this needs to be per project, since each project might be connected to a different gitlab instance or use a differnt gitlab user

mc1arke commented:
SonarSource no longer seem to document the properties used for configuring each Pull Request provider. I'll see if I can confirm what they've used for Gitlab, but am happy you use the names you've suggested given they seem sensible.

mc1arke commented:
The changes for Github and Bitbucket decoration have now been merged into the target pr-decoration branch. This has caused some conflicts against your changes in common files, and also needs updates in your changes for some constants that @4n4n4s created for the properties shared between your 2 changes. Could you rebase your MR and resolve these conflicts please?

tisoft commented:
Rebased, will now cleanup and add tests

mfolnovic commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/GitlabServerPullRequestDecorator.java:
```diff
@@ -0,0 +1,388 @@
+            LOGGER.info(String.format("Commits in MR: %s ", commits.stream().map(Commit::getId).collect(Collectors.joining(", "))));
+            List<CommitDiscussion> commitDiscussions = new ArrayList<>();
+            for (Commit commit : commits) {
+                getCommitDiscussions(projectURL + String.format("/repository/commits/%s/discussions", commit.getId()), headers, deleteCommentsEnabled)
```
what's the difference between creating discussions on MR's commits VS directly on MR?

there's an API for creating discussions or MR directly: https://docs.gitlab.com/ee/api/discussions.html#create-new-merge-request-thread

tisoft replied:
I am unsure about the difference between the 2 APIs. The one I'm using also creates a discussion, that can be replied to.

I will also try the MR discussion API out, to see if that has any differences. But I want to rebase, cleanup and test the current code before I do that.

mfolnovic replied:
cool :+1: 

what I would check is if discussion on commit shows up as thread in MR:

![image](https://user-images.githubusercontent.com/20919/71890273-354d0a80-3144-11ea-8ce3-8546df6f6b0d.png)

and if you turn on:
![image](https://user-images.githubusercontent.com/20919/71890353-5f9ec800-3144-11ea-97de-8a1ba881d52f.png)

then you should see:
![image](https://user-images.githubusercontent.com/20919/71890330-53b30600-3144-11ea-8f82-5bc5f8bbd9ea.png)

tisoft replied:
I don't see that with the commit comments.

I will play around with the MR discussions.

tisoft replied:
I have played around with MR discussions. code and result can be seen here: https://gitlab.com/tisoft/sonarqube-community-branch-plugin-gitlab-test/merge_requests/1

Resolving of discussions is currently not done by the plugin, it will delete all discussions and recreate them at each run.

What do you think?

mfolnovic replied:
nice, looks awesome, well done! :100: :muscle: 

may I suggest adding a link to Sonar issue and/or Sonar rule? :angel:

I would also try creating discussions first, and then creating MR comment with summary, so that would also be at the end (before user comments, of course). What do you think?

not sure if it's something on my end (tried also incognito), but images aren't loading for me:
![image](https://user-images.githubusercontent.com/20919/71916552-7530e380-317e-11ea-988c-ed385b424ba0.png)

other than that, looks really nice, thanks for trying this out! :blush:

tisoft replied:
Personally I like having the summary on top, so that I have a quick overview and don't need to do endless scrolling before I can get to that.

Images are loaded from the github master branch of this repo, and since this is not yet merged into master, there are no images yet. The contents of the summary and issue comments is shared between all the decoration implementations, I don't want to change that as part as of this PR, but I agree that a link back to sonar would be helpful. Will do a followup PR as soon as everything else is settled.

![image](https://user-images.githubusercontent.com/139041/71917358-24ba8580-3180-11ea-9049-5ce3236b25b3.png)

If you click on the failed pipeline marker on the righrt and then on SonarQube, you are send to the sonar instance. (which will not work since its running on localhost on my machine :laughing:)

tisoft commented on src/main/java/com/github/mc1arke/sonarqube/plugin/CommunityBranchPlugin.java:
```diff
@@ -67,8 +72,131 @@ public void load(CoreExtension.Context context) {
+                        .type(PropertyType.STRING)
+                        .build(),
+
+                PropertyDefinition.builder("sonar.pullrequest.summary.comment.enabled")
```
cnahged the prefix to com.github.mc1arke.sonarqube.plugin.branch.pullrequest for the properties as its done in the bitbucket implementation

mc1arke commented:
Is this ready for final review and merge? If so, please remove the 'WIP' from the title.

tisoft commented:
Its now ready from my point of view. I have changed it to use merge request discussions instead of file comments, since MR discussions look better in the gitlab merge request views.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/GitlabServerPullRequestDecorator.java:
```diff
@@ -0,0 +1,354 @@
+    }
+
+    private <X> List<X> getPagedList(String commitDiscussionURL, Map<String, String> headers, boolean sendRequest, TypeReference<List<X>> typeRef) throws IOException {
+         HttpGet httpGet = new HttpGet(commitDiscussionURL);
```
Nitpick: incorrect indent (one additional space)

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/GitlabServerPullRequestDecorator.java:
```diff
@@ -0,0 +1,354 @@
+                String.format("%s must be specified in the project configuration", propertyName)));
+    }
+
+    private static String getNextUrl(HttpResponse httpResponse) {
```
Nitpick: should this not return `Optional<String>` rather than using `null` and requiring calling code to perform a null check?

tisoft commented:
fixed your comments in a new commit for easier review.

feel free to squash all commits into a single one.

don-vip commented:
@tisoft how does the repository slug configuration works? I only see the parameter available in global administration settings and not in the project settings. How can I analyse several GitLab repositories with the plugin?

tisoft commented:
I have it in the project settings:

![Screenshot_2020-01-09 General Settings - test](https://user-images.githubusercontent.com/139041/72099976-09877b80-3322-11ea-8d50-3bb14a5bbe96.png)

don-vip commented:
@tisoft  thanks! I wasn't looking in the right place

don-vip commented:
@tisoft des it work with SonarQube 8.0 or do we need 8.1? I can't trigger the MR decoration on my projects. I'm running a Maven build from GitLab CI with 
`mvn clean verify sonar:sonar $MVN_OPTS -Dsonar.pullrequest.key=$CI_MERGE_REQUEST_IID -Dsonar.pullrequest.branch=$CI_MERGE_REQUEST_SOURCE_BRANCH_NAME -Dsonar.pullrequest.base=$CI_MERGE_REQUEST_TARGET_BRANCH_NAME`

The branch analysis is performed, but nothing happens concerning the decoration. I have configured GitLab host and acces token in global settings, project slug in project settings, and sonar.pullrequest.* parameters on the command line. Am I missing something? In DEBUG level I can just see this in ce.log:

```
2020.01.10 15:34:22 INFO  ce[AW-QF4KJnUt6s8owbZ4T][o.s.c.t.s.ComputationStepExecutor] Index analysis | status=SUCCESS | time=16ms
2020.01.10 15:34:22 INFO  ce[AW-QF4KJnUt6s8owbZ4T][o.s.c.t.s.ComputationStepExecutor] Send issue notifications | status=SUCCESS | time=0ms
2020.01.10 15:34:22 INFO  ce[AW-QF4KJnUt6s8owbZ4T][o.s.c.t.s.ComputationStepExecutor] Publish task results | status=SUCCESS | time=0ms
2020.01.10 15:34:22 INFO  ce[AW-QF4KJnUt6s8owbZ4T][o.s.c.t.s.ComputationStepExecutor] Trigger refresh of Portfolios and Applications | status=SUCCESS | time=0ms
2020.01.10 15:34:22 INFO  ce[AW-QF4KJnUt6s8owbZ4T][o.s.c.t.p.a.p.PostProjectAnalysisTasksExecutor] Webhooks | globalWebhooks=0 | projectWebhooks=0 | status=SUCCESS | time=0ms
2020.01.10 15:34:22 DEBUG ce[AW-QF4KJnUt6s8owbZ4T][c.g.m.s.p.c.p.PullRequestPostAnalysisTask] found 3 pull request decorators
2020.01.10 15:34:22 INFO  ce[AW-QF4KJnUt6s8owbZ4T][o.s.c.t.p.a.p.PostProjectAnalysisTasksExecutor] Pull Request Decoration | status=SUCCESS | time=0ms
```
